snapcraft/_store.py

```diff
@@ -272,12 +274,88 @@ def download(snap_name, channel, download_path, arch):
+    store = storeapi.StoreClient()
+    # Get data for the gating snap
+    with _requires_login():
+        snap_data = store.cpi.get_package(snap_name, 'stable', None)
+    snap_id = snap_data['snap_id']
+    validations = store.get_validations(snap_id)
+
+    table_data = []
+    for v in validations:
+        table_data.append([v['approved-snap-name'],
+                           v['approved-snap-revision']])
+    tabulated = tabulate(table_data, headers=['Name', 'Approved'],
+                         tablefmt="plain")
+    print(tabulated)
+
+
+def validate(snap_name, validations, revoke=False, key=None):
+    """Generate, sign and upload validation assertions."""
+
+    # Check validations format
+    validation_re = re.compile('^[^=]+=[0-9]+$')
+    invalids = [v for v in validations if not validation_re.match(v)]
+    if invalids:
+        for v in invalids:
+            logger.error('Invalid validation request "{}", format must be'
+                         ' name=revision'.format(v))
+        raise RuntimeError()
+
+    store = storeapi.StoreClient()
+
+    # Get data for the gating snap
+    with _requires_login():
+        snap_data = store.cpi.get_package(snap_name, 'stable', None)
+    release = str(snap_data['release'][0])
+    snap_id = snap_data['snap_id']
+
+    # Need the ID of the logged in user.
+    with _requires_login():
+        account_info = store.get_account_information()
+    authority_id = account_info['account_id']
+
+    # Then, for each requested validation, generate assertion
+    for validation in validations:
+        gated_name, rev = validation.split('=', 1)
+        approved_data = store.cpi.get_package(gated_name, 'stable', None)
+        assertion = {
+            'type': 'validation',
+            'authority-id': authority_id,
+            'series': release,
+            'snap-id': snap_id,
+            'approved-snap-id': approved_data['snap_id'],
+            'approved-snap-revision': rev,
+            'timestamp': datetime.datetime.utcnow().isoformat() + 'Z',
+            'revoked': "false"
+        }
+        if revoke:
+            assertion['revoked'] = "true"
+
+        cmdline = ['snap', 'sign']
+        if key:
+            cmdline += ['-k', key]
+        snap_sign = subprocess.Popen(
+            cmdline, stdin=subprocess.PIPE, stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
+        data = json.dumps(assertion).encode('utf8')
+        logger.info('Signing validation {}'.format(validation))
+        assertion, err = snap_sign.communicate(input=data)
+        if snap_sign.returncode != 0:
+            err = err.decode('ascii', errors='replace')
+            raise RuntimeError('Error signing assertion: {!s}'.format(err))
+
+        # Save assertion to a properly named file
+        fname = '{}-{}-r{}.assertion'.format(snap_name, gated_name, rev)
+        with open(fname, 'wb') as f:
+            f.write(assertion)
+
+        store.push_validation(snap_id, assertion)
```

Comment:
I split it in 3, there is a largish chunk in the loop that's hard to split without writing functions with many args, which are a separate readability problem. Let me know if you like the current split or want me to split finer.